src/comm/XPUPair.h

```diff
@@ -0,0 +1,246 @@
+#pragma once
+
+#include <ATen/Context.h>
+
+namespace at::xpu {
+
+// define null_type
+struct null_type {};
+
+// null_type comparisons
+inline bool operator==(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator>=(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator<=(const null_type&, const null_type&) {
+  return true;
+}
+
+inline bool operator!=(const null_type&, const null_type&) {
+  return false;
+}
+
+inline bool operator<(const null_type&, const null_type&) {
+  return false;
+}
+
+inline bool operator>(const null_type&, const null_type&) {
+  return false;
+}
+
+// forward declaration for tuple
+template <
+    class T0 = null_type,
+    class T1 = null_type,
+    class T2 = null_type,
+    class T3 = null_type,
+    class T4 = null_type,
+    class T5 = null_type,
+    class T6 = null_type,
+    class T7 = null_type,
+    class T8 = null_type,
+    class T9 = null_type>
+class tuple;
+
+template <typename T1, typename T2>
+inline void swap(T1& a, T2& b) {
+  T1 temp = a;
+  a = b;
+  b = temp;
+}
```

Comment:
Why is this reimplementing swap and not using `std::swap`? This doesn't work correctly if `T1` or `T2` have their custom swap specialization.